src/services/bunker.py

```diff
@@ -1,15 +1,350 @@
+            logger.info({"msg": "Bunker ON. High midterm slashing penalty"})
+            return True
+        if self._is_abnormal_cl_rebase(cdtp, frame_cl_rebase):
+            logger.info({"msg": "Bunker ON. Abnormal CL rebase"})
+            return True
+
+    def _get_cl_rebase_for_frame(self, cdtp: CommonDataToProcess) -> Gwei:
+        logger.info({"msg": "Getting CL rebase for frame"})
+        ref_lido_balance, ref_withdrawal_vault_balance = self._calculate_lido_balance_with_vault(
+            cdtp.ref_blockstamp, [v.validator for v in cdtp.ref_lido_validators], cdtp.ref_blockstamp
+        )
+        # Make a static call of 'handleOracleReport' without EL rewards
+        # to simulate report and get total pool ether after that
+        args = {
+            "_reportTimestamp": cdtp.ref_timestamp,
+            "_timeElapsed": cdtp.seconds_elapsed_since_last_report,
+            "_clValidators": len(cdtp.ref_lido_validators),
+            "_clBalance": self.w3.to_wei(ref_lido_balance, 'gwei'),
+            "_withdrawalVaultBalance": ref_withdrawal_vault_balance,
+            "_elRewardsVaultBalance": 0,
+            "_lastFinalizableRequestId": 0,
+            "_simulatedShareRate": 0,
+        }
+        before_report_total_pooled_ether = self.w3.lido_contracts.lido.functions.totalSupply().call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        after_report_total_pooled_ether, _ = self.w3.lido_contracts.lido.functions.handleOracleReport(**args).call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        logger.info({
+            "msg": "Simulate 'handleOracleReport' contract function call",
+            "call_args": args,
+            "before_call": before_report_total_pooled_ether,
+            "after_call": after_report_total_pooled_ether,
+        })
+
+        frame_cl_rebase = self.w3.from_wei(after_report_total_pooled_ether - before_report_total_pooled_ether, 'gwei')
+        logger.info({"msg": f"Simulated CL rebase for frame: {frame_cl_rebase} Gwei"})
+
+        return frame_cl_rebase
+
+    def _is_high_midterm_slashing_penalty(self, cdtp: CommonDataToProcess, cl_rebase: Gwei) -> bool:
+        logger.info({"msg": "Detecting high midterm slashing penalty"})
+        all_slashed_validators = self._not_withdrawn_slashed_validators(cdtp.ref_all_validators, cdtp.ref_epoch)
+        lido_slashed_validators: List[Validator] = [
+            v.validator for v in self.w3.lido_validators.filter_lido_validators(cdtp.lido_keys, all_slashed_validators)
+        ]
+        logger.info({"msg": f"Slashed - all: {len(all_slashed_validators)} lido: {len(lido_slashed_validators)}"})
+
+        # If no one lido in current slashed validators - no need to bunker
+        if len(lido_slashed_validators) == 0:
+            return False
+
+        # Fill per epoch buckets by possible slashed epochs
+        per_epoch_buckets: dict[int, List[Validator]] = {}
```

Comment:
naming is controversial, I agree, but in this step of reading it is more important to us that this "something" per epoch, because it's just an intermediate state of data. I expand a code comment above

Reply:
done